internal/dot/graph.go

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dot
+
+// Node represents a single node in a graph.
+type Node struct {
+	Type     string
+	Name     string
+	Optional bool
+	Group    string
+}
+
+// Edge connects a node parameter to a node result.
+type Edge struct {
+	Param  *Node
+	Result *Node
+}
+
+// Graph represents the DOT-format graph in a Container.
+type Graph struct {
+	Edges []*Edge
+}
+
+// Add adds the edges in node n into dg.
+func (dg *Graph) Add(params []*Node, results []*Node) {
+	edges := make([]*Edge, len(params)*len(results))
+
+	for i, param := range params {
+		for j, result := range results {
+			edges[i*len(results)+j] = &Edge{Param: param, Result: result}
```

Comment:
You can do this without the index calculation math if you use the
`make([]foo, length, capacity)` form of `make`.

```go
edges := make([]*Edge, 0, len(params)*len(results))
```

This creates a slice that has room for that many items but it's still empty
(so `edges[0]` will panic).

Then it's just a matter of appending in the right order, which in this case is
iteration order:

```go
for _, param := range params {
    for _, result := range results {
        edges = append(edges, &Edge{Param: param, Result: result})
    }
}
```

That said, it looks like creating this slice is unnecessary because you can
just append to `dg.Edges` directly. Is there a reason not to do that?

Reply:
I was attempting to avoid the unnecessary size re-allocation in the `dg.Edges` slice here so I used a separate `edge` slice and appended it on after. I think that the best approach would be to take your
```
edges := make([]*Edge, 0, len(params)*len(results))
```
suggestion and use `append` afterwards. 

What are your thoughts? Do you think the `make` allocation is unnecessary/overkill?